src/ADMM.jl

```diff
@@ -16,22 +16,22 @@ mutable struct ADMM{matT,opT,ropT,vecT,rvecT,preconT} <: AbstractLinearSolver
   u::Vector{vecT}
   # other parameters
   precon::preconT
-  ρ::rvecT
-  iterations::Int64
-  iterationsInner::Int64
+  ρ::rvecT # TODO: Switch all these vectors to Tuple
+  iterations::Integer
```

Comment:
Got it, thank you for the detailed explanations. I changed `Integer` back to `Int64`, I hear you that it does not make much of a difference, even though 64bit is probably an overkill for all use cases. 

My main incentive was to ensure that `Float32` inputs won't be promoted to 64bit, as I want to be able to run my reconstructions with single precision. I replaced the `Real` now with `T` and introduced a type `TC <: Union{T, Complex{T}}` to cover complex and real valued cases. I also replaced the `vecT` and `rvecT` by `Vector{TC}` and `Vector{T}`, respectively. Does that make sense to you and will that break any of your other code?

Reply:
> My main incentive was to ensure that Float32 inputs won't be promoted to 64bit, as I want to be able to run my reconstructions with single precision. 

That is a very good reason I am in favor of allowing Float32/Float64 in a more consistent fashion (using type parameters). We were a little bit lazy in this respect.

Reply:
>  I also replaced the vecT and rvecT by Vector{TC} and Vector{T}, respectively. Does that make sense to you and will that break any of your other code?

That breaks code, see
https://github.com/tknopp/RegularizedLeastSquares.jl/commit/6074824b46c91d30a239dda5d38437a3f6c1338b

we definitely don't want to loose CUDA support. We need to think about what is the proper solution there.

Reply:
Hmm, too bad. What are the requirements for CUDA here? I was trying to steer away from definitions like `vecT`, next to `opT`. To my understanding, and please tell me if I am mistaken, they both have T in their name, but there is nothing that ensure that `T` is the same in both. And it would be nice to ensure that all operators / vectors / matrices use the same floating type +/- complex/real versions wherever appropriate, wouldn't it?

Reply:
What about this:
```
struct Test{T,vecT<:AbstractVector{T}} 
           a::vecT
           b::T
       end
```
just as an abstract example of how this can be better solved.

Reply:
Sounds good. Do I understand correctly, that CUDA just requires an `AbstractVector` instead of `Vector`?

Reply:
> What are the requirements for CUDA here?

CUDA has a type `CuArray{T,D}` and both `Array{T,D}` and `CuArray{T,D}` are `AbstractArray{T,D}`. At function boundaries we would thus use `AbstractArray{T,D}`. But in struct this does not work since we don't want to have abstract members.

> To my understanding, and please tell me if I am mistaken, they both have T in their name, but there is nothing that ensure that T is the same in both.

Correct, `vecT` is just a name. You can also name that `G` or whatever. But my trick above achieves what we want.

Reply:
>  Do I understand correctly, that CUDA just requires an AbstractVector 

Kind of. But its not a CUDA requirement. Its more that we cannot assume everything to be a `Vector/Array`. So `AbstractVector/AbstractArray` will generalize enough such that we can support CUDA (or possibly also OpenCL).

Reply:
I implemented it the way you suggested. Do you mind taking a look?